### PR TITLE
[SPARK-12968][SQL] Implement command to set current database

### PR DESCRIPTION
**Pull request by viirya**
JIRA: https://issues.apache.org/jira/browse/SPARK-12968

Implement command to set current database.

### CODE REVIEW

**Comment by rxin** [comment]
cc @hvanhovell for review.

**Comment by viirya** [comment]
retest this please.

File: sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/CliSuite.scala [hunk 1]
```diff
@@ -183,7 +183,7 @@ class CliSuite extends SparkFunSuite with BeforeAndAfterAll with Logging {
       "CREATE DATABASE hive_test_db;"
         -> "OK",
       "USE hive_test_db;"
-        -> "OK",
+        -> "",
```
Comment (hvanhovell) [inline sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/CliSuite.scala]: Maybe a dumb idea: we could return Ok...

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveQuerySuite.scala [hunk 2]
```diff
@@ -1262,6 +1263,21 @@ class HiveQuerySuite extends HiveComparisonTest with BeforeAndAfter {
 
   }
 
+  test("use database") {
+    val currentDatabase = sql("select current_database()").first().getString(0)
+
+    sql("CREATE DATABASE hive_test_db")
+    sql("USE hive_test_db")
+    assert("hive_test_db" == sql("select current_database()").first().getString(0))
```
Comment (hvanhovell) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveQuerySuite.scala]: The ```current_database()``` command is currently added in the HiveContext. We should move this to the ```SQLContext```.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/commands.scala [hunk 3]
```diff
@@ -408,3 +408,13 @@ case class DescribeFunction(
     }
   }
 }
+
+case class SetDatabaseCommand(databaseName: String) extends RunnableCommand {
```
Comment (hvanhovell) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/commands.scala]: We are introducing the concept of a database change in Catalyst. Shouldn't we also have ```SetDatabaseCommand``` in Catalyst?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Catalog.scala [hunk 4]
```diff
@@ -46,6 +46,10 @@ trait Catalog {
 
   def lookupRelation(tableIdent: TableIdentifier, alias: Option[String] = None): LogicalPlan
 
+  def setCurrentDatabase(databaseName: String): Unit = {
+    throw new UnsupportedOperationException
```
Comment (hvanhovell) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Catalog.scala]: Why create a dummy implementation here? This is not consistent with the way Catalog is currently written. The other option creates more overhead though.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/client/ClientInterface.scala [hunk 5]
```diff
@@ -109,6 +109,9 @@ private[hive] trait ClientInterface {
   /** Returns the name of the active database. */
   def currentDatabase: String
```
Comment (hvanhovell) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/client/ClientInterface.scala]: Shouldn't we also have a ```currentDatabase``` function in the Catalog?

Comment (rxin) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/client/ClientInterface.scala]: Yes we should - the only thing is that the simple catalog in non-Hive doesn't support databases yet.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Catalog.scala (hunk 4 above)
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Catalog.scala]: I think not all catalog support database concept. Actually, inherited catalog can choose to implement it or not.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/client/ClientInterface.scala (hunk 5 above)
Comment (viirya) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/client/ClientInterface.scala]: Yeah. I think we don't need to address database support for all catalogs in this PR?

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveQuerySuite.scala (hunk 2 above)
Comment (viirya) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveQuerySuite.scala]: Do we already have database support in `SQLContext`?

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/client/ClientInterface.scala (hunk 5 above)
Comment (hvanhovell) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/client/ClientInterface.scala]: I would argue that when an interface allows us to set the current database, it should also allow us read it.

Comment (viirya) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/client/ClientInterface.scala]: I can add a default implementation.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/commands.scala (hunk 3 above)
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/commands.scala]: If you mean should we move parsing `TOK_SWITCHDATABASE` to CatalystQl, as currently I see SQLContext use SparkQl not CatalystQl, this should not be a problem now. I have no strong opinion about this. We can move it now or when Catalyst supports database. How do you think?

**Comment by rxin** [comment]
@viirya can you remove ur last commit? I can merge this. I'm redesigning catalog right now and we don't need to add that no-op commit here.

**Comment by viirya** [comment]
@rxin ok. I will update this later (about 30 mins) when I am back to desk.

**Comment by viirya** [comment]
@rxin I've reverted last commit.

**Comment by rxin** [comment]
Thanks - I've merged this.

**Pull request closed without merging**